osu.Game/Screens/Edit/EditorCommandHandler.cs

```diff
@@ -0,0 +1,312 @@
+                apply(command);
+
+            editorBeatmap.EndChange();
+
+            if (changeHandler != null)
+                changeHandler.SuppressStateChange = false;
+        }
+
+        private void historyChanged()
+        {
+            CanUndo.Value = undoStack.Count > 0;
+            CanRedo.Value = redoStack.Count > 0;
+        }
+
+        private void apply(IEditorCommand command)
+        {
+            command.Apply();
+            CommandApplied?.Invoke(command);
+        }
+
+        private void record(IEditorCommand command)
+        {
+            var reverse = command.CreateUndo();
+
+            currentTransaction.Add(reverse);
+        }
+
+        private readonly struct Transaction
+        {
+            public readonly bool IsChangeHandlerTransaction;
+
+            public Transaction()
+            {
+                undoCommands = new List<IEditorCommand>();
+            }
+
+            public Transaction(bool isChangeHandlerTransaction = false)
+            {
+                IsChangeHandlerTransaction = isChangeHandlerTransaction;
+                undoCommands = new List<IEditorCommand>();
+            }
+
+            private Transaction(List<IEditorCommand> undoCommands, bool isChangeHandlerTransaction = false)
+            {
+                IsChangeHandlerTransaction = isChangeHandlerTransaction;
+                this.undoCommands = undoCommands;
+            }
+
+            private readonly List<IEditorCommand> undoCommands;
+
+            /// <summary>
+            /// The commands to undo the given transaction.
+            /// Stored in reverse order of original commands to match execution order when undoing.
+            /// </summary>
+            public IReadOnlyList<IEditorCommand> UndoCommands => undoCommands;
+
+            public void Add(IEditorCommand command)
+            {
+                if (command is IMergeableCommand mergeable)
+                {
+                    for (int i = 0; i < undoCommands.Count; i++)
+                    {
+                        // Since the commands are stored in reverse order (to match execution order when undoing), the
+                        // command we're inserting is treated as the previous command relative to the current one.
+                        if (mergeable.MergeWithNext(nextCommand: undoCommands[i], out var merged))
+                        {
+                            undoCommands[i] = merged;
+
+                            // Since currently there's only one command that a given command can be merged with, we can
+                            // stop iterating through the list once we've found a match.
+                            return;
+                        }
+                    }
+                }
+
+                undoCommands.Insert(0, command);
+            }
+
+            public Transaction Reverse()
+            {
+                var commands = UndoCommands.Reverse().Select(command => command.CreateUndo()).ToList();
```

Comment:
The command handler is a loose port of the command handler I wrote for osucad, which was built to support multiplayer editing. Multiplayer editing is something that I want to try and push for in lazer too eventually which is why I've tried to not introduce anything that would be incompatible with that.

The basic premise here is that if you have multiple users, while undo will generally always stay the same action, redo won't. If you redo a change, the value might have already been changed by someone else, so redo has to go back to the current state. Generating a new command based on the current state ensures that this always works.
```
let X = 0
- User A:     X = 2
- User B:     X = 5
- User A:     undo -> X = 0
- User A:     redo -> X = 5 // redo needs to revert to 5. if we had reused the original command, x would be 2 now
```

Reply:
Also just coming back to this, I find the example above disputable. I claim two things:

- As fun as it is to nerd out over such things, the resolution of conflicts of this sort is *secondary* as nobody *wants* to be touching parts of a map 100% concurrently with another user. Which is to say you could probably do whatever and get away with shrugging your arms and going "well then don't do that then" because it's such a rare use case to have to 100% correctly CRDT these or whatever that it's not worth the mental overhead everywhere else.
- I also think that the premise of "shared user undo history" is likely to be more confusing than it's worth and if anything, undo should be local to *each user individually*. What does that mean in terms of a conflict? See point 1 wherein it probably doesn't matter that much because any resolution you go with is weird enough to the users that it essentially doesn't matter. Like in the example above, wherein I find the notion of shared history and "reverting to 5" as just not intuitive.

Reply:
It is indeed fun to nerd out about such things so I'll try to give some insight into how I'm doing these things in osucad to hopefully explain my reasonings for these things a bit (though I already have a draft for a discussion post where I go into much more detail).

> * nobody _wants_ to be touching parts of a map 100% concurrently with another user

I have encountered this plenty of times in osucad so saying that it won't happen isn't really correct in my opinion. However this isn't so much meant to account for completely concurrent editing of the same object and more to deal with a different case:
The undo system (most fo the multiplayer command handling actually) is largely inspired by figma & what they share about it in their dev blog. The redo correction is primarily to ensure the following requirement is met:
![image](https://github.com/user-attachments/assets/e3a512c8-6917-440b-b5f6-cbbd729a1d85)
*[source article](https://www.figma.com/blog/multiplayer-editing-in-figma/)*

> * because it's such a rare use case to have to 100% correctly CRDT these or whatever that it's not worth the mental overhead everywhere else.

CRDTs are more trouble than it's worth. Luckily you don't need them when you have an authorative server which acts as a source of truth for the order in which operations were applied, as CRDTs are used for the case when you don't know the order. The way I am doing conflict resolution is much simpler, and works very well. I made sure there's no operations that do list modifications (apart from sorted lists since you can figure out the order locally there), so the vast majority of commands just do simple property updates. For those you just need to discard incoming operations until the latest change has been ack'd by the server and that's about it.
On a side note, concurrent list modifications (primarily dealing insert & delete) are genuinely the most unintuitive problem that I ever encountered which is why I'm always just sending the entire list with the updates applied to it and treat it like a regular property change. The two cases where I have regular list modifications in commands are slider paths & hitsounds. I don't think this is solvable without something like vector clocks so it's probably best to just avoid the problem alltogether.

> * I also think that the premise of "shared user undo history" is likely to be more confusing than it's worth and if anything, undo should be local to _each user individually_. What does that mean in terms of a conflict?

I never mentioned a shared user undo history, undo is still local and I never planned for it to be shared. It would just need to be corrected for redo to match user expectations. For conflict (other user makes changes since you hit undo, and then you redo after) the way I did it the redo command just override any changes. There's no 100% correct way of handling that case so I just went for the one that has the least amount of complexity.

> * Like in the example above, wherein I find the notion of shared history and "reverting to 5" as just not intuitive.

I think if I redo 3 times, then redo 3 times the intuitive thing would be to go back to the state I had before I pressed undo, and not back to "the state as if whatever the changes other people made to the object didn't happen".

Reply:
> The undo system (most fo the multiplayer command handling actually) is largely inspired by figma & what they share about it in their dev blog. The redo correction is primarily to ensure the following requirement is met:
> ![image](https://private-user-images.githubusercontent.com/8039761/376566627-e3a512c8-6917-440b-b5f6-cbbd729a1d85.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg5OTA4MDIsIm5iZiI6MTcyODk5MDUwMiwicGF0aCI6Ii84MDM5NzYxLzM3NjU2NjYyNy1lM2E1MTJjOC02OTE3LTQ0MGItYjVmNi1jYmJkNzI5YTFkODUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTVUMTEwODIyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDIzMDAzYmIzYTQ4YjBlOTI3ZmI3ZWQ3NWJkZGY3YmJmMTA2N2YzY2E3ZDVmMGUyOGI0N2ZlZTRmMTIzZmM0NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.-9XbbBfpJjQ7NpSGQbgmA54u8A-_ZjzVu59npDes1bo)

Point taken, but I'm still not sure you *need* all of the complexity incurred here to implement it this way. If I'm understanding the problem correctly (which I admit I might not), it's a question of applying the command "absolutely" contrary to "relative to current ambient state". Which is simple if you _write the command that way_, and populate its state upon the original submission of the command. (See framework transforms, by the way, which do this.)

Reply:
Fair. I have to admit a lot of the complexity in this PR comes from me and oli both just throwing our ideas in here with too little regard about what is actually needed at this point and what isn't. We'll come back with a much smaller prototype that deals just with a minimum use case.

> Which is simple if you _write the command that way_, and populate its state upon the original submission of the command.

That seems like a reasonable approach, I'll try tinkering with it.

Reply:
I'm not sure I understand the above conversation.

[If we have an authoritative server...](https://github.com/ppy/osu/pull/30255#discussion_r1800917675)
[And the sequence...](https://github.com/ppy/osu/pull/30255#discussion_r1797989588)

```
let X = 0
- User A:     X = 2
- User B:     X = 5
- User A:     undo -> X = 0
- User A:     redo -> X = 5
```

Then isn't step 3 wrong? It should undo to X=2 if you have a shared timeline.

If you partially-separated timelines, which I don't think are a bad idea in principle, then I think of it somewhat like a git history:

```
          User A
             |
             v
X = 0 ---> X = 2 ------> X = 2/5
   \                  /
    -----> X = 5 -----
             ^
             |
          User B
```

- Both A and B start with X=0 as revision 1.
- The change from A arrives to the server first, it becomes "mainline". (In reality this would follow the same steps as below, but I'm simplifying).
- The change from B arrives to the server second.
- Because mainline is no longer on revision 1, we first check for validity. "Is the state of X at revision 2 equal to the state at revision 1?"
- If yes - apply the change to mainline.
- If no - tell the user that the object has been changed and ask them what to do about it - a two-way merge.

So the structure I envision in that scenario is something like:
```csharp
class ChangeX
{
    int OldValue;
    int NewValue;

    void Apply(Beatmap target)
    {
        ASSERT(target.X == OldValue);
        target.X = NewValue;
    }
    
    void Undo(Beatmap target)
    {
        ASSERT(target.X == NewValue);
        target.X = OldValue;
    }
    
    bool CheckForConflict(Beatmap other)
    {
        return other.X != OldValue;
    } 
}
```

So I'm not seeing what the value in "CreateUndo" is really.

Reply:
> Then isn't step 3 wrong? It should undo to X=2 if you have a shared timeline.

Marvin did mention we don't intend to have a shared timeline

> I never mentioned a shared user undo history, undo is still local and I never planned for it to be shared.

I also think a shared undo timeline would not be a good user experience. You might hit undo, see nothing change on your screen, but undo a change of someone else on another part of the map.

Reply:
> * Both A and B start with X=0 as revision 1.
> * The change from A arrives to the server first, it becomes "mainline". (In reality this would follow the same steps as below, but I'm simplifying).
> * The change from B arrives to the server second.
> * Because mainline is no longer on revision 1, we first check for validity. "Is the state of X still equal to the state at revision 1?"
> * If yes - apply the change to mainline.
> * If no - tell the user that the object has been changed and ask them what to do about it - a two-way merge.

I think presenting the user with conflict resolution prompts is not very good ux, as well as causing the issue that any continuous operation suddenly needs to be interruptible to present such prompt (also, which user will get to decide how to resolve the conflict?). Most other tools/libraries that do this go for a "last write wins" approach (i.e. figma, microsoft fluid-framework), since it's generally simpler & more intuitive for the end user. For undo, there's no 100% correct way of doing things, as pretty much all approaches have reasons in favor and against them. However I found that when trying out various approaches, just ignoring other changes when undoing is what felt the most natural, undo doing anything other than undoing *my* change feels kinda weird. 
I don't have access to figma's source code but from observation & their blog I believe the undo system they use is, conceptually, the same that I have.
Having shared undo also comes with the issue of having one extra thing to have to do conflict resolution for, what if 2 people undo at the same time before their undo-commands reach each other?

Reply:
Sure, I agree with that. And just going back over this thread I saw the pseudocode in https://github.com/ppy/osu/pull/30255/files#r1800917675, which is pretty much what I expect.